### ISSUE TITLE
stb_truetype: stbtt_InitFont missing parameter for size of "data"

### ISSUE THREAD

**Issue by bitc** [nothings/stb#690]
Here is the signature of stbtt_InitFont:

    int stbtt_InitFont(stbtt_fontinfo *info, const unsigned char *data, int offset);

It is missing a parameter for the size of the data buffer. Otherwise how can stb know when it reaches the end of the data?

It seems that stb just assumes that it always receives the correct amount of data. I'm pretty sure that this can lead to buffer overflow, reading of invalid memory, segfaults, and even potential security issues.

I am aware that this library is intended to be used for games, where the font file is known to be valid and not malicious, but I believe that all software should be written to correctly validate all input. And there are certainly plenty of legitimate scenarios where a font file becomes accidentally corrupted or truncated, and stb should properly handle that.

Thank you

**Comment by bitc** [comment]
Also `stbtt_GetNumberOfFonts` and `stbtt_GetFontOffsetForIndex` are vulnerable

**Comment by nothings** [comment]
Yes, if you want a security guarantee, validate the file before calling stb_truetype functions.

**Comment by bitc** [comment]
ok thank you for your response

**Comment by nothings** [comment]
BTW, I agree that it is an oversight that the length is not specified... I don't know what I was thinking!

But changing it now would break everyone, so...

**Comment by bitc** [comment]
Btw, I haven't looked too deeply at the code, but I believe that fixing this shouldn't be too hard. Just add a "length" field to `stbtt_fontinfo`. And then modify all the lines of code that access the buffer to first check that the offset they are accessing is less than "length".

This could be added without breaking backwards incompatibility by adding a new function:

    void stbtt_FontSetDataLength(stbtt_fontinfo *info, int length);

Users of the library would be required to call this immediately after calling `stbtt_InitFont`.

Alternatively, maybe it would be possible to add an additional `int length` argument to the end of `stbtt_InitFont`. I think this might be possible to do without breaking backwards incompatibility using C varargs or some other technique.

(You could even add a warning message that is printed if you try to call one of the font functions before calling `stbtt_FontSetDataLength`, this way, people using the library would get notified that they should update their code. But this warning might be too invasive depending on your tastes)

**Comment by rygorous** [comment]
There's a lot of sideband functions that take the raw unsigned char* pointer, also without a size, so unfortunately this is easier said than done and can't easily be fixed without breaking changes to half the API.
